gadget/install/partition.go

```diff
@@ -306,54 +301,49 @@ func removeCreatedPartitions(gadgetRoot string, lv *gadget.LaidOutVolume, dl *ga
 	return nil
 }
 
-func partitionsWithRolesAndContent(lv *gadget.LaidOutVolume, dl *gadget.OnDiskVolume, roles []string) []onDiskAndLaidoutStructure {
+func partitionsWithRolesAndContent(lv *gadget.LaidOutVolume, dl *gadget.OnDiskVolume, roles []string) []gadget.LaidOutStructure {
 	roleForOffset := map[quantity.Offset]*gadget.LaidOutStructure{}
-	for idx, gs := range lv.LaidOutStructure {
-		if gs.Role() != "" {
-			roleForOffset[gs.StartOffset] = &lv.LaidOutStructure[idx]
+	for idx, los := range lv.LaidOutStructure {
+		if los.Role() != "" {
+			roleForOffset[los.StartOffset] = &lv.LaidOutStructure[idx]
 		}
 	}
 
-	var odloStructures []onDiskAndLaidoutStructure
+	var loStructures []gadget.LaidOutStructure
 	for _, part := range dl.Structure {
-		// Create per-iter var from loop variable as we store the pointer in odls
-		part := part
-		gs := roleForOffset[part.StartOffset]
-		if gs == nil || gs.Role() == "" || !strutil.ListContains(roles, gs.Role()) {
+		laidOut := roleForOffset[part.StartOffset]
+		if laidOut == nil || laidOut.Role() == "" || !strutil.ListContains(roles, laidOut.Role()) {
 			continue
 		}
-		// now that we have a match, set the laid out structure such
-		// that the fields reflect what was declared in the gadget, the
-		// on-disk-structure already has the right size as read from the
-		// partition table
-		odls := onDiskAndLaidoutStructure{
-			onDisk:  &part,
-			laidOut: gs,
-		}
-		odloStructures = append(odloStructures, odls)
-	}
-	return odloStructures
+		// now that we have a match, set the on-disk-structure structure
+		// in the laid out structure. on-disk-structure already has the
+		// right size as read from the partition table
+		// TODO fill in construction of LaidOutStructure instead?
+		laidOut.OnDiskStructure = part
```

Comment:
this function and applyOnDiskStructureToLaidOut do this kind of things, is there some way to share/unify some code across them?

Reply:
They are called in slightly different contexts, the first from factory reset and the second from the install API. I think they could be merged but it would need some additional refactoring. The next step after this PR would be to try to build LaidOutStructure only when we have all data available, I think that when we do that these will be merged as part of that change.